tests/kernel/timer/timer_monotonic/src/main.c

```diff
@@ -8,21 +8,139 @@
 #include <tc_util.h>
 #include <ztest.h>
 
-int test_frequency(void)
+static struct k_timer test_timer;
+static struct k_timer test_periodtimer;
+
+u32_t start_time, stop_time, cycles_spent, nanoseconds_spent,
+	   time_spent[10], cycles_cost[10], pctlist[10];
+int count = 0;
+
+
+static void callback_period(struct k_timer *timer)
+{
+	u32_t pct;
+
+	stop_time = k_cycle_get_32();
+	cycles_spent = stop_time - start_time;
+	nanoseconds_spent = (u32_t)k_cyc_to_ns_floor64(cycles_spent);
+	pct = (u64_t)cycles_spent * 100000U / sys_clock_hw_cycles_per_sec();
+	start_time = stop_time;
+	pctlist[count] = pct;
+	time_spent[count] = nanoseconds_spent;
+	cycles_cost[count] = cycles_spent;
+
+	pct =0;
+
+	count++;
+	if (count >= 10) {
+		k_timer_stop(timer);
+	}
+}
+
+static void callback_duration(struct k_timer *timer)
+{
+	stop_time = k_cycle_get_32();
+}
+
+int test_periodtimer_frequency(void)
 {
-	u32_t start, end, delta, pct;
+	int i;
+
+	TC_PRINT("Testing system tick frequency\n");
+
+    k_timer_init(&test_periodtimer, callback_period, NULL);
+    start_time = k_cycle_get_32();
+    k_timer_start(&test_periodtimer, 1, 1);
+    k_sleep(1000);
+    k_timer_stop(&test_periodtimer);
+    for(i = 0; i < 10; i++)
+    {
+       printk("delta: %u expected cycles: %u "
+			   "expected time(nanosec): %u "
+			   "cost time(nanosec): %u  %u%%\n",
+			   cycles_cost[i], sys_clock_hw_cycles_per_sec()/1000,
+			   1000000 ,time_spent[i], pctlist[i]);
+
+      /* Heuristic: if we're more than 10% off, throw an error */
+       if (pctlist[i] < 90 || pctlist[i] > 110) {
+		   TC_PRINT("Clock calibration is way off!\n");
+		   return -1;
+		   break;
+	    }
+
+	     time_spent[i] = 0;
+         cycles_cost[i] = 0;
+         pctlist[i] = 0;
+      }
+
+      start_time = 0;
+      stop_time = 0;
+      cycles_spent = 0;
+      nanoseconds_spent = 0;
+
+      return 0;
+}
+
+int test_oneshottimer_frequency(void)
+{
+	u32_t delta, pct;
+
+	TC_PRINT("Testing system tick frequency\n");
+
+    k_timer_init(&test_timer, callback_duration, NULL);
+    start_time = k_cycle_get_32();
+
+    k_timer_start(&test_timer, 1, 0);
+    k_timer_status_sync(&test_timer);
+    delta = stop_time - start_time;
+    nanoseconds_spent = (u32_t)k_cyc_to_ns_floor64(delta);
+    k_timer_stop(&test_timer);
+	pct = (u64_t)delta * 100000U / sys_clock_hw_cycles_per_sec();
+
+	printk("delta: %u expected cycles: %u "
+			"expected time(nanosec): %u "
+			"cost time(nanosec): %u  %u%%\n",
+			delta, sys_clock_hw_cycles_per_sec()/1000,
+			1000000, nanoseconds_spent, pct);
+
+	start_time = 0;
+    stop_time = 0;
+    nanoseconds_spent = 0;
+	/* Heuristic: if we're more than 10% off, throw an error */
+	if (pct < 90 || pct > 110) {
+		TC_PRINT("Clock calibration is way off!\n");
```

Comment:
Phrasing: this error message got copied from somewhere else.  "Way off" in English means "very wrong", but this will fire on some configurations (e.g. 100 Hz ticks) with just a one-tick error in the result.

Reply:
For the below codes, you are right and I was copy it from the original case . In fact, I was develop the 3 test new cases for 1 millisecond based on the original case structure to test for  “k_sleep()”, “one shot timer” and “period timer”. 
So is the original case also have problem ? and if possible , could you please share some suggestion how to update it? Thanks in advance. 

==============code =========================
			if (pct < 90 || pct > 110) {
			     TC_PRINT("Clock calibration is way off!\n");
                      }
==========================================

Reply:
Just change the string to "Clock calibration is off"